string message
int32 threshold